### PR TITLE
Fix measurement of Row and Column on iOS inside of UIStackView

### PR DESCRIPTION
**Pull request by squarejesse**
UIStackView uses UIView.intrinsicContentSize on its subviews, but our Row and Column implementations don't have a natural implementation of this API. In particular, if any subview wraps its contents (so its height depends on its available width), there's no single intrinsicContentSize().

We work around this by following some online advice: invalidate the intrinsic size when the bounds change, and then use those bounds when computing the new intrinsic size.

This is more fragile and stateful than I'd like, but it seems like a common practice for working around UIStackView's API.

---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: redwood-lazylayout-composeui/src/test/snapshots/images/app.cash.redwood.layout.composeui_ComposeUiLazyListTest_RTL_testIntrinsicContentSizeWhenSubviewsRequireScrolling.png
Comment (squarejesse) [inline redwood-lazylayout-composeui/src/test/snapshots/images/app.cash.redwood.layout.composeui_ComposeUiLazyListTest_RTL_testIntrinsicContentSizeWhenSubviewsRequireScrolling.png:1]: More ComposeUI layouts that are broken. This is a problem, but it’s not new and it’s not urgent.

File: redwood-snapshot-testing/src/iosMain/kotlin/app/cash/redwood/snapshot/testing/UIViewTestWidgetFactory.kt [hunk 1]
```diff
@@ -120,8 +120,8 @@ class UIViewSimpleColumn : SimpleColumn<UIView> {
 
   override val value = UIStackView(CGRectZero.readValue()).apply {
     this.axis = UILayoutConstraintAxisVertical
-    this.alignment = UIStackViewAlignmentLeading
-    this.distribution = UIStackViewDistributionFill
+    this.alignment = UIStackViewAlignmentFill
+    this.distribution = UIStackViewDistributionEqualSpacing
```
Comment (squarejesse) [inline redwood-snapshot-testing/src/iosMain/kotlin/app/cash/redwood/snapshot/testing/UIViewTestWidgetFactory.kt]: I had better luck with these ones

File: redwood-snapshot-testing/src/iosMain/kotlin/app/cash/redwood/snapshot/testing/UIViewTestWidgetFactory.kt [hunk 2]
```diff
@@ -144,8 +144,8 @@ class UIViewScrollWrapper : ScrollWrapper<UIView> {
       scrollView.addSubview(value)
       value.translatesAutoresizingMaskIntoConstraints = false
       value.leadingAnchor.constraintEqualToAnchor(scrollView.leadingAnchor).active = true
+      value.widthAnchor.constraintEqualToAnchor(scrollView.widthAnchor).active = true
```
Comment (squarejesse) [inline redwood-snapshot-testing/src/iosMain/kotlin/app/cash/redwood/snapshot/testing/UIViewTestWidgetFactory.kt]: It’s way too cute, but UIScrollView scrolls horizontally if you bind `leading` and `trailing`, but doesn’t scroll horizontally if you bind `leading` and `width`.

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt [hunk 3]
```diff
@@ -45,7 +49,48 @@ internal class YogaUIView : UIScrollView(cValue { CGRectZero }), UIScrollViewDel
...
+      }
+    }
+
+    super.setBounds(bounds)
+  }
+
+  override fun invalidateIntrinsicContentSize() {
+    super.invalidateIntrinsicContentSize()
+    this.widthForIntrinsicSize = UIViewNoIntrinsicMetric
+    this.heightForIntrinsicSize = UIViewNoIntrinsicMetric
```
Comment (squarejesse) [inline redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt]: I don’t have a test case to prove that this override is necessary. I might try to make one? That might be tricky

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt [hunk 4]
```diff
@@ -45,7 +49,48 @@ internal class YogaUIView : UIScrollView(cValue { CGRectZero }), UIScrollViewDel
...
+   *
+   *  1. Making [intrinsicContentSize] depend on the mostly-recently applied bounds
+   *  2. Invalidating it each time the bounds change
+   *
+   * This will result in an additional layout pass when the parent view uses [intrinsicContentSize].
+   */
+  override fun setBounds(bounds: CValue<CGRect>) {
+    val newWidth = bounds.useContents { size.width }
+    val newHeight = bounds.useContents { size.height }
+    if (isColumn()) {
```
Comment (squarejesse) [inline redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt]: (Maybe instead I wanna check if the width mode is Fill here)

File: redwood-layout-uiview/RedwoodLayoutUIViewTests/__Snapshots__/UIViewFlexContainerTestHost/testIntrinsicContentSizeWhenSubviewsRequireScrolling.1.png
Comment (squarejesse) [inline redwood-layout-uiview/RedwoodLayoutUIViewTests/__Snapshots__/UIViewFlexContainerTestHost/testIntrinsicContentSizeWhenSubviewsRequireScrolling.1.png:1]: Without the fix, this thing’s intrinsicContentSize is is 10,000 pixels wide and 20 pixels tall

testIntrinsicContentSizeWhenSubviewsRequireScrolling 1

File: redwood-layout-uiview/RedwoodLayoutUIViewTests/__Snapshots__/UIViewFlexContainerTestHost/testIntrinsicContentSizeWhenSubviewsWrap.1.png
Comment (squarejesse) [inline redwood-layout-uiview/RedwoodLayoutUIViewTests/__Snapshots__/UIViewFlexContainerTestHost/testIntrinsicContentSizeWhenSubviewsWrap.1.png:1]: Similarly, without the fix each subview is super wide and not very tall

testIntrinsicContentSizeWhenSubviewsWrap 1

**Review by edahle — APPROVED** [review]
🔥 🔥

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt [hunk 5]
```diff
@@ -45,7 +49,54 @@ internal class YogaUIView : UIScrollView(cValue { CGRectZero }), UIScrollViewDel
...
-  override fun intrinsicContentSize(): CValue<CGSize> = calculateLayout()
+  /**
+   * The intrinsic size is broken by design if any subview's height depends on its width (or
+   * vice-versa). For example, if a subview is UILabel that wraps, we need to know how wide the
+   * label is before we can compute that label's height.
+   *
+   * We work around this by:
+   *
+   *  1. Making [intrinsicContentSize] depend on the mostly-recently applied bounds
+   *  2. Invalidating it each time the bounds change
```
Comment (colinrtwhite) [inline redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt:60]: This work around sounds good to me, though @edahle do you know how a `UITextView` and other views that depend on knowing the width to compute the height provide a reasonable `intrinsicContentSize`? Maybe private APIs of some kind?

**Review by colinrtwhite — APPROVED** [review]
Thanks for fixing this! 🙏🏻

File: redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt (hunk 5 above)
Comment (squarejesse) [reply redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt:60]: UILabel has a feature called [preferredMaxLayoutWidth](https://developer.apple.com/documentation/uikit/uilabel/1620534-preferredmaxlayoutwidth) that you can use for this.

**Pull request merged**